internal_task_handlers.go

```diff
@@ -724,45 +729,123 @@ func newActivityTaskHandler(activities []activity,
 }
 
 type cadenceInvoker struct {
-	identity      string
-	service       m.TChanWorkflowService
-	taskToken     []byte
-	cancelHandler func()
-	retryPolicy   backoff.RetryPolicy
+	sync.Mutex
+	identity              string
+	service               m.TChanWorkflowService
+	taskToken             []byte
+	cancelHandler         func()
+	retryPolicy           backoff.RetryPolicy
+	heartBeatTimeoutInSec int32 // The heart beat interval configured for this activity.
+	hbBatchEndTimer       *time.Timer
+	lastProgressToReport  *[]byte
+	isInBatchingWindow    bool // Whether we started a batch of operations that need to be reported in the cycle. This gets started on a user call.
+	closeCh               chan struct{}
 }
 
 func (i *cadenceInvoker) Heartbeat(details []byte) error {
+	i.Lock()
+	defer i.Unlock()
+
+	if i.isInBatchingWindow {
+		// If we have started batching window, keep track of last reported progress.
+		i.lastProgressToReport = &details
+		return nil
+	}
+
+	nonTransientError, err := i.internalHeartBeat(details)
```

Comment:
Can this call happen on the background go routine? Can we simplify this method to just start the background go routine?